runtime/src/global_request_scheduler_mts.c

```diff
@@ -0,0 +1,333 @@
+	if (priority_queue_delete_nolock(destination_queue, top_mgrq) == -1)
+		panic("Tried to delete an MGRQ from the Global runqueue, but was not present");
+	// debuglog("Removed the MGRQ from the Global runqueue");
+
+	if (priority_queue_length_nolock(top_mgrq->sandbox_requests) > 0) {
+		priority_queue_enqueue_nolock(destination_queue, top_mgrq);
+		// debuglog("Added the MGRQ back to the Global runqueue to Heapify");
+	}
+
+done:
+	LOCK_UNLOCK(&global_lock);
+	return rc;
+}
+
+/**
+ * Peek at the priority of the highest priority task without having to take the lock
+ * Because this is a min-heap PQ, the highest priority is the lowest 64-bit integer
+ * This is used to store an absolute deadline
+ * @returns value of highest priority value in queue or ULONG_MAX if empty
+ */
+static uint64_t
+global_request_scheduler_mts_peek(void)
+{
+	uint64_t val = priority_queue_peek(global_request_scheduler_mts_guaranteed);
+	if (val == UINT64_MAX) val = priority_queue_peek(global_request_scheduler_mts_default);
+
+	return val;
+}
+
+
+uint64_t
+global_request_scheduler_mts_guaranteed_peek(void)
+{
+	return priority_queue_peek(global_request_scheduler_mts_guaranteed);
+}
+
+uint64_t
+global_request_scheduler_mts_default_peek(void)
+{
+	return priority_queue_peek(global_request_scheduler_mts_default);
+}
+
+
+/**
+ * Initializes the variant and registers against the polymorphic interface
+ */
+void
+global_request_scheduler_mts_initialize()
+{
+	global_request_scheduler_mts_guaranteed = priority_queue_initialize(RUNTIME_RUNQUEUE_SIZE, false,
+	                                                                            module_request_queue_get_priority);
+	global_request_scheduler_mts_default    = priority_queue_initialize(RUNTIME_RUNQUEUE_SIZE, false,
+                                                                                 module_request_queue_get_priority);
+
+	global_module_timeout_queue = priority_queue_initialize(RUNTIME_RUNQUEUE_SIZE, false,
+	                                                        module_timeout_get_priority);
+
+	LOCK_INIT(&global_lock);
+
+	struct global_request_scheduler_config config = {
+		.add_fn                  = global_request_scheduler_mts_add,
+		.remove_fn               = global_request_scheduler_mts_remove,
+		.remove_if_earlier_fn    = global_request_scheduler_mts_remove_if_earlier,
+		.remove_with_mt_class_fn = global_request_scheduler_mts_remove_with_mt_class,
+		.peek_fn                 = global_request_scheduler_mts_peek
+	};
+
+	global_request_scheduler_initialize(&config);
+}
+
+void
+global_request_scheduler_mts_free()
+{
+	priority_queue_free(global_request_scheduler_mts_guaranteed);
+	priority_queue_free(global_request_scheduler_mts_default);
+	priority_queue_free(global_module_timeout_queue);
+}
+
+
+void
+global_timeout_queue_add(struct module *module)
+{
+	struct module_timeout *mt = malloc(sizeof(struct module_timeout));
```

Comment:
You an avoid this allocation eventually by simply having the timer queue internal node be a part of the module. For now, the malloc is fine.

Reply:
Nice catch! Noted and done!